sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,210 @@
+                     pika.URLParameters(os.environ['AMQP']))
+        return pika.URLParameters(os.environ['AMQP'])
+
+    def on_conn_close(self, reply_code, reply_text):
+        """Called when RabbitMQ has been connected to.
+
+        :param int reply_code: The code for the disconnect
+        :param str reply_text: The disconnect reason
+
+        """
+        LOGGER.warning('RabbitMQ has disconnected (%s): %s',
+                       reply_code, reply_text)
+        AMQPMixin.connection = None
+        self.amqp_ready.clear()
+        AMQPMixin.amqp_state = AMQP_STATES['disconnected']
+        self._connect_to_rabbitmq()
+
+    @gen.coroutine
+    def on_conn_open(self, connection):
+        """Called when RabbitMQ has been connected to.
+
+        :param pika.connection.Connection connection: The pika connection
+
+        """
+        LOGGER.info('RabbitMQ has created a connection')
+        AMQPMixin.connection = connection
+        AMQPMixin.connection.add_on_close_callback(self.on_conn_close)
+        yield self._create_new_channel()
+        self.amqp_ready.set()
+        LOGGER.info('RabbitMQ has connected')
+
+    def on_channel_open(self, channel):
+        """Called when the RabbitMQ accepts the channel open request.
+
+        :param pika.channel.Channel channel: The channel opened with RabbitMQ
+
+        """
+        LOGGER.info('Channel %i is opened for communication with RabbitMQ',
+                    channel.channel_number)
+        AMQPMixin.channel = channel
+        AMQPMixin.amqp_state = AMQP_STATES['connected']
+        self.amqp_channel_ready.set()
+
+    def on_channel_close(self, channel):
+        """Called when the RabbitMQ accepts the channel close request.
+
+        :param pika.channel.Channel channel: The channel closed with RabbitMQ
+
+        """
+        LOGGER.info('Channel %i is closed for communication with RabbitMQ',
+                    channel.channel_number)
+        AMQPMixin.channel = None
+        self.amqp_channel_ready.clear()
+        AMQPMixin.amqp_state = AMQP_STATES['disconnected']
+
+    def reset_amqp_state(self):
+        """ Clear the amqp_ready event, clear the amqp_channel_ready event and
+        set the AMQP_STATE to connecting.
+
+        """
+        AMQPMixin.amqp_state = AMQP_STATES['connecting']
+        self.amqp_ready.clear()
+        self.amqp_channel_ready.clear()
+
+    @gen.coroutine
+    def prepare(self):
+        """Prepare the handler, ensuring RabbitMQ is connected or start a new
+        connection attempt.
+
+        If a new connection needs ot be created, this will wait until the
+        connection is established.
+
+        Waiting for a new connection will timeout in WAIT_SECONDS and return a
+        504.
+
+        """
+        maybe_future = super(AMQPMixin, self).prepare()
+        if maybe_future:
+            yield maybe_future
+
+        if self._finished:
+            return
+
+        if not self.connection and self.amqp_state != 'connecting':
```

Comment:
AFAICT, `amqp_state` will always be one of `{'CONNECTED', 'CONNECTING', 'DISCONNECTED'}`.  Should this be `self.amqp_state != AMQP_STATES['connecting']`?  The same comment applies to line 208 below.  We can probably jettison the `AMQP_STATES` dict altogether since it really doesn't buy us anything.  That is, unless I am missing something.